src/content/dataSubmissions/SubmittedData.tsx

```diff
@@ -177,53 +262,159 @@ const SubmittedData: FC<Props> = ({ submissionId, submissionName }) => {
     setLoading(false);
   };
 
-  const handleFilterChange = (filters: FilterForm) => {
-    filterRef.current = filters;
-    tableRef.current?.setPage(0, true);
+  const handleFilterChange = useCallback(
+    (filters: FilterForm) => {
+      setSelectedItems([]);
+      filterRef.current = filters;
+      tableRef.current?.setPage(0, true);
+    },
+    [setSelectedItems, filterRef.current, tableRef.current]
+  );
+
+  const handleToggleRow = useCallback(
+    (nodeIds: string[]) => {
+      setSelectedItems((prev) => {
+        const newSelected = new Set(prev);
+        nodeIds.forEach((id) => {
+          if (newSelected.has(id)) {
+            newSelected.delete(id);
+          } else {
+            newSelected.add(id);
+          }
+        });
+        return Array.from(newSelected);
+      });
+    },
+    [setSelectedItems]
+  );
+
+  const handleToggleAll = useCallback(async () => {
+    // NOTE: Force immediate update of selectedItems for the current page
+    flushSync(() => {
+      setSelectedItems(data?.map((node) => node.nodeID) || []);
+    });
+
+    // If all rows are already visible, no need to fetch data
+    if (data?.length === totalData) {
+      isFetchingAllData.current = false;
+      return;
+    }
+
+    const { data: d, error } = await getSubmissionNodes({
```

Comment:
This should be wrapped in a try/catch otherwise will cause uncaught exception.

Reply:
Thanks for catching this! Fixed in 8564282